pkg/storage/fs/ocis/upload.go

```diff
@@ -401,16 +425,67 @@ func (upload *fileUpload) FinishUpload(ctx context.Context) (err error) {
 	}
 	targetPath := upload.fs.lu.toInternalPath(n.ID)
 
+	sublog := appctx.GetLogger(upload.ctx).With().Interface("info", upload.info).Str("binPath", upload.binPath).Str("targetPath", targetPath).Logger()
+
+	// calculate the checksum of the written bytes
+	// TODO only calculate in sync if checksum was sent?
+	sha1h := sha1.New()
+	md5h := md5.New()
+	adler32h := adler32.New()
+
+	f, err := os.Open(upload.binPath)
+	if err != nil {
+		sublog.Err(err).Msg("ocisfs: could not open file for checksumming")
+		// we can continue if no oc checksum header is set
+	}
+	defer f.Close()
+
+	r1 := io.TeeReader(f, sha1h)
+	r2 := io.TeeReader(r1, md5h)
+
+	if _, err := io.Copy(adler32h, r2); err != nil {
+		sublog.Err(err).Msg("ocisfs: could not copy bytes for checksumming")
+	}
+
+	// compare if they match the sent checksum
+	// TODO the tus checksum extension would do this on every chunk, but I currently don't see an easy way to pass in the requested checksum. for now we do it in FinishUpload which is also called for chunked uploads
```

Comment:
Would the client send the checksum only with the last chunk or with all of those?

Reply:
it would send the checksum for a PATCH request. without the chunking extension that is the complete file. with the chunking extension it would be the checksum of the chunk.

Reply:
But when `FinishUpload` is called, `binPath` would have the complete file, right? In that case, the last chunk request would need to have have the checksum of the whole file, otherwise the asserts would fail.

Reply:
The `metadata["checksum"]` is set in `InitiateUpload()`, which will store it in the info file. So the checksum (if one was sent) is available in FinishUpload(). Even if it was not requested, we still need to store it in the metadata, which is why we have to split checking them from persisting them. The latter of which has to happen always.